### PR TITLE
ENH: stats: exponential, logarithm, and reciprocal of a random variable

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Toward gh-15928

#### What does this implement/fix?
SciPy distributions like `lognorm`, `loglaplace`, and `loguniform` are examples of distributions underlying random variables whose logarithm is distributed according to a more fundamental distribution. To generalize this idea and avoid naming inconsistencies[^1], this PR adds a class that represents the distribution underlying a strictly monotonic function of a random variable, and a funtion `exp` that takes the exponential of a random variable. For example, if `X = stats.Normal()`, `Y = exp(X)` is lognormally distributed. While I was at it, I went ahead and overrode the `__rtruediv__` operator, so if `X` is gamma-distributed, `Y = 1 / X` is inverse gamma distributed.

#### Additional information
This class makes it easy to define other transformations of random variables including `log` and `ContinuousDistribution.__pow__`[^2]. `log` and `exp` probably shouldn't be public functions of the `stats` subpackage - maybe we need a new namespace? Or make these static methods (or methods?) of a superclass `ProbabilityDistribution`?

[^1]: Confusingly, `loggamma` represents the distribution of a random variable that *is the logarithm* of a gamma-distributed random variable (rather than "whose logarithm is a gamma-distributed random variable").
[^2]: `__pow__` will require a little more work because taking even powers of a distribution whose support includes negative numbers is not an injective operation.

### CODE REVIEW

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:5307]:
```suggestion

```

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:5256]:
```suggestion
```

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:5260]:
```suggestion
    def _logpdf_formula(self, x, *args, **params):
```
Something to consider in the future for all transformed distributions - don't override the `_dispatch` function, just the `_formula`.

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:5224]:
```suggestion
    Given a random variable :math:`X`; a strictly increasing function
```

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:5222]:
```suggestion
    r"""Distribution underlying a strictly increasing function of a random variable
```

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:5302]:
```suggestion

```

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:5324]:
```suggestion
    If :math:`X` is to be the natural logarithm of :math:`Y`, then we
    must take :math:`Y` to be the natural exponential of :math:`X`.
```

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:5328]:
```suggestion
    To demonstrate that ``X`` represents the logarithm of ``Y``,
```

**Comment by fancidev** [comment]
It occurred to me that a potential design question worth considering is whether/how to support transformation of a distribution instance (random variable) versus transformation of a distribution class.

For example, `Mixture([x, y])` takes random variables as input and produces a random variable as output. So do `log(x)` and `exp(x)`. In comparison, `ContinuousDistribution.from_rv_continuous` produces a distribution class as output.

Do you foresee support for transformation of a distribution class? For example, something like `LogLaplace = exp(Laplace)` to get a *class* derived from `ContinuousDistribution`? That could affect the naming decisions and also the implementation.

**Comment by mdhaber** [comment]
I did think about that recently. 

With respect to `exp` in particular, I think it is more natural for an `exp` function to be used to 
"exponentiate $X$" 
than 
"create the distribution underlying the exponential of a random variable that follows a ___ distribution"

Also, when we get into composable transformations (like shifting/scaling + truncation or vice versa, see gh21752), I think it is a better interface to accept the parameters required for each transformation separately, especially because the order of the transformations matters. While we could conceivably transform distribution classes with particular parameters, it seems a little strange. 

```python3
truncate(2*Normal() + 1, -3, 5)
```
seems better than
```python3
# not natural to shift/scale a class, and we would have
# to get clever to make this possible
truncate(2*Normal + 1, -3, 5)()  
```
and definitely better than
```python3
truncate(shift_scale(Normal))(lb = -3, ub=5, loc=1, scale=2)
```

For mixtures, I like the idea of tossing a bunch of objects of different sizes (weights) into an urn, picking one, then observing it (object based). But I can also see mixture as defining a distribution by doing a weighted sum of PDFs (class based). But again, what does the interface look like if you want to apply transformations to the random variables (object-based) / distributions (class-based)?

`ContinuousDistribution.from_rv_continuous` (or whatever we call it) is different because it's not a transformation. It always comes before transformations, never after. We *could* require the user to pass the parameters into the `from_rv_continuous` function and have it return an object (instead of a class). But I tend not to like accepting arguments that simply get passed through unmodified to something else (like *args in minimize, etc).

**Comment by fancidev** [comment]
After more thinking it seems to me that variable transform and distribution transform *could* be supported with the same signature. For example,

```python
truncate(Normal(), left=1, right=2)
```

would produce a random variable, while

```python
truncate(Normal, left=Parameter('a'), right=Parameter('b'))
```

could produce a distribution. I'm not saying this "overload" is necessarily ideal, but at least it seems "future-proof" to implement  random variable transformation first without worrying about distribution transform.

And I think it's probably a good idea to put the relevant functions into a sub-namespace,  say `scipy.stats.transforms`. Those functions could include shift, scale, log, exp, reciprocal, mixture, compound, sum, ratio, etc.

I feel that's not a bad idea.

**Comment by josef-pkt** [comment]
Sounds again that you don't like parameter estimation.

parameter need to be mutable for any kind of `fit` function. 

e.g. my ancient version of a transformed distribution class
https://github.com/statsmodels/statsmodels/blob/main/statsmodels/sandbox/distributions/transformed.py#L143

(statsmodels is mostly into regression and estimation, so this will not change the need to keep our own distributions.)

**Comment by mdhaber** [comment]
> Sounds again that you don't like parameter estimation.

Let's not make this personal, please.

But I do like parameter estimation. Here is a small selection of my work improving `scipy.stats` in this area:
- https://github.com/scipy/scipy/pull/15436
- https://github.com/scipy/scipy/issues/11782
- https://github.com/scipy/scipy/pull/16967
- https://github.com/scipy/scipy/pull/17141
- https://github.com/scipy/scipy/pull/19475
 
and my original motivation for creating the `scipy.differentiate` sub-package was to compute Fisher information matrices.

I will take the comment to mean "Have you considered parameter estimation? How will that work?"

Here is an example of fitting working with this PR as-is, without mutable random variables. It can be made much faster when we decide what to do with mutability, which is orthogonal to this issue. (At some point I was asked to leave them immutable for now due to interface concerns, but I plan to tackle that at some point.)

<details>

```python3
from scipy import stats

import numpy as np
import matplotlib.pyplot as plt
from scipy.stats import TruncatedDistribution, Normal
from scipy import optimize

rng = np.random.default_rng(834589345982345)

def get_truncnorm(x):
    u, s, lb, ub = x
    X = Normal()
    Y = s * TruncatedDistribution(X, lb=lb, ub=ub) + u
    return Y

def fit(data, guess):
    def f(x, data):
        X = get_truncnorm(x)
        lb, ub = X.support()
        i = (lb < data) & (data < ub)
        return -X.logpdf(data[i]).sum()

    def constraint(x, data):
        u, s, lb, ub = x
        high = (ub*s + u) - data
        low = data - (lb*s + u)
        return np.concatenate([high, low])

    lb = [-np.inf, 1e-2, -np.inf, -np.inf]
    ub = [np.inf, np.inf, np.inf, np.inf]

    bounds = optimize.Bounds(lb, ub)
    res = optimize.minimize(f, x0=guess, bounds=bounds,
                            constraints={'type': 'ineq', 'fun': lambda x: constraint(x, data)},
                            args=(data,), method='slsqp')
    return res.x

guess = [0, 1, -1, 1]
Y =  get_truncnorm([1, 2, -2, 2])
data = Y.sample(100, rng=rng)

params = fit(data, guess)
Y = get_truncnorm(params)
x = np.linspace(-5, 5, 300)
plt.plot(x, Y.pdf(x), '--', label='fit')
plt.hist(data, density=True, bins=30, label='histogram')
plt.legend()
plt.show()

# Doesn't work
# params = stats.truncnorm.fit(data, -1, 1, loc=0, scale=1)
# Y = get_truncnorm(params)
# x = np.linspace(-5, 5, 300)
# plt.plot(x, Y.pdf(x), '--', label='fit')
# plt.hist(data, density=True, bins=30, label='histogram')
# plt.legend()
# plt.show()
```

</details>

image

I couldn't get `scipy.stats.truncnorm` to fit this. Also note that this PR, unlike `truncnorm`, will allow you to hold the truncation points fixed. https://github.com/scipy/scipy/issues/15928 was one of the reasons for allowing composable transformations like this.

**Comment by mdhaber** [comment]
Oops, looks like I got this mixed up with the `TruncatedDistribution` PR. But the same thoughts apply here.

**Comment by josef-pkt** [comment]
MLE for support parameters is nonstandard.

I meant parameters of the underlying distribution.
e.g. standard parameterization of log-normal uses mean and variance of the underlying normal distribution (in contrast to the loc-scale parameterization of scipy.stats lognorm)

e.g Log-t: y = exp(x), x ~ t(loc, scale, df)
where z = a + b y, for additional loc-shape of Log-t distribution.
This would have parameter (a, b, loc, scale, df) to estimate.

Anyway:
This is not personal, this is focusing on different application direction between statsmodels and where scipy stats distributions are going.

**Comment by mdhaber** [comment]
IIUC, that's just easier. I don't have a t distribution in this branch, but here we recover the location and scale of an underlying normal distribution.

```python3

import numpy as np
from scipy.stats import exp, Normal
from scipy import optimize

rng = np.random.default_rng(834589345982345)

def get_lognorm(x):
    u, s = x
    return exp(s*Normal() + u)

def fit(data, guess):
    def f(x, data):
        X = get_lognorm(x)
        return -X.logpdf(data).sum()

    lb = [-np.inf, 1e-3]
    ub = [np.inf, np.inf]

    bounds = optimize.Bounds(lb, ub)
    res = optimize.minimize(f, x0=guess, bounds=bounds,
                            args=(data,))
    return res.x

guess = [0, 1]
X = 2*Normal() + 1
data = np.exp(X.sample(1000, rng=rng))
params = fit(data, guess)
print(params)
# [0.96778275 1.96778275] vs the original [1, 2]
```

**Comment by josef-pkt** [comment]
The idea of adding transformation based distributions is good.
I wrote a lot of code like that for statsmodels that unfortunately never became public, supported code.

Thanks for `validation_policy`.
We will see how it works when statsmodels uses the new distribution framework. (I guess it will not happen soon.)

(your example is still not really relevant for the missing `args`. The (log)pdf of the new class will not allow for params arguments. But it matches the current new distribution pattern.)

**Comment by mdhaber** [comment]
> your example is still not really relevant for the missing args. The (log)pdf of the new class will not allow for params arguments. But it matches the current new distribution pattern

The example was only intended to address the concerns about the ability to fit parameters of the underlying distribution. If it does not, can you try rephrasing your concern? 

> After more thinking it seems to me that variable transform and distribution transform could be supported with the same signature.

@fancidev IIUC you mean that there should be two -- and preferably only two -- obvious way to do it. Is that right?

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:1840]: For now, raise if 0 is in the support.

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:1841]:
```suggestion
                                          dh=lambda u: 1 / u ** 2, increasing=False) * other
```

**Comment by fancidev** [comment]
> IIUC you mean that there should be two -- and preferably only two -- obvious way to do it. Is that right?

I guess the confusion here is what precise is "it" in the above. I think there are three levels of transforms implicitly being considered here:

a. transform of a random variable
b. transform of a distribution
c. transform of a distribution family

Those are three different (but related) tasks.

The PR deals with (a), and implicitly (b). It does not deal with (c) directly, though you suggest it's possible to create a custom wrapper (e.g. via implementing `__new__`). I'm not totally convinced custom wrappers qualify for "an obvious way" if we treat distribution families as having "first-class" support.

By the way, if I understand correctly, in the work done so far with respect to the new distribution infrastructure, "random variable" and "distribution" are implicitly interchangeable. This looks fine so far, but I suppose the distinction would become important as soon as binary transforms are introduced, where dependence between operands comes into play.

**Comment by mdhaber** [comment]
They are not, but the confusion is noted. I planned to write a tutorial as soon as we have a few of these things merged and there is something to write about. Shall we discuss it then?

File: scipy/stats/_distribution_infrastructure.py
Comment (steppi) [inline scipy/stats/_distribution_infrastructure.py:5248]: We need to overload  `__repr__`.  Right now it just uses the __repr__ for `dist`

```python
import scipy.stats as stats

X = stats.Normal(mu=0, sigma=1)
Y = stats.exp(Y)
print(Y)
# Normal(mu=0.0, sigma=1.0)
```

It's not obvious what to do though. Since what is now called `IMFTransformedDistribution` is intended to be private, we can get away with passing a function or data needed for  `__repr__` as an argument to `__init__`. 

Something like this is likely the best we can do, since the `reprs` for `g`, `h`, etc. will be opaque things like `<function <lambda> at 0x714ea2b96700>`.  Even then, we'd be confounded by inconsistent naming conventions for transformed distributions, like you pointed out for `lognormal` vs `loggamma`, so even for specific transforms, there might not be a clean thing to do.

For now, let's just think about what to do for the specific cases in this PR.

[Inverse Distribution](https://en.wikipedia.org/wiki/Inverse_distribution) is pretty standard for the reciprocal of a random variable, so in that case, i think it makes sense to make the `repr` `Inv<something>`.

For `exp`, `"exp(Normal(mu=0.0, sigma=1.0)"` is confusing in my opinion. 

Something like

`"exp(X): X ~ Normal(mu=0.0, sigma=1.0)"`

is more clear, but a bit verbose, and it's not clear that X is being used more like a free variable, and doesn't refer to any particular X.

`"exp(*): * ~ Normal(mu=0.0, sigma=1.0)"`  perhaps clears up the issue that we don't mean a particular X or whatever, but is confusing in its own right.

Any thoughts?

File: scipy/stats/_distribution_infrastructure.py
Comment (steppi) [reply scipy/stats/_distribution_infrastructure.py]: There's still an argument to `__init__` for whether the transform is increasing or decreasing and the reciprocal implemented here involves a decreasing transform. Was the suggested change above from before you supported the decreasing case and implemented the reciprocal?

I think `MonotonicTransformedDistribution` would work well here.

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [reply scipy/stats/_distribution_infrastructure.py:5248]: Good catch. To put it in perspective, there is a very low bar to meet. For a few decades, frozen instances of `rv_continuous` have looked like:

```
import scipy
scipy.stats.norm(1, 2)
# <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x796e4c735090>
```

regardless of the underlying distribution family, and apparently nobody has complained. Of course, I'd like to do better; I just thought it was worth mentioning where we're coming from.

Re: 

> For exp, "exp(Normal(mu=0.0, sigma=1.0))" is confusing in my opinion.

Why? That is actually a valid `repr` (i.e. that is actually code to reproduce the object), and it's pretty readable. Since the standards for naming the distributions underlying transformations of random variables are not very rigid, I don't think we're going to be able to do something like `LogNormal(mu=0.0, sigma=1.0)` in general, so I think showing how the random variable itself is transformed is the best thing to do.

`inv(Normal(mu=0.0, sigma=1.0))` works, but I didn't plan on exposing an `inv` function, and we're probably going to add powers, after this and the `abs` PR are done, so maybe it should be `(Normal(mu=0.0, sigma=1.0))**(-1)`?

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [reply scipy/stats/_distribution_infrastructure.py]: I think it was from before. Sure, `MonotonicTransformedDistribution` is fine with me.

File: scipy/stats/_distribution_infrastructure.py
Comment (steppi) [reply scipy/stats/_distribution_infrastructure.py:5248]: I guess `exp(Normal(mu=0.0, sigma=1.0))` is fine. I was thinking in terms of distributions, not in terms of Python objects, it makes it look like we're taking `exp` of a distribution. `1 / Normal(mu=0.0, sigma=1.0)` seems fine for division. I guess the `repr` needs to not just be a string representation, but a way to generate an instance from a string right? In which case those seem like the best options.

Having an opaque thing analogous to `# <scipy.stats._distn_infrastructure.rv_continuous_frozen object at 0x796e4c735090>
` is fine too, if not ideal, and could be improved later. Having it be the `repr` for a different distribution is wading into bug territory though imo, and should be fixed in some way.

File: scipy/stats/_distribution_infrastructure.py
Comment (steppi) [inline scipy/stats/_distribution_infrastructure.py:5222]: If we think about it, `ShiftedScaleDistribution` is a special case of what is now called `IMFTransformedDistribution`. Not something for now, but just to think about for the future. To reduce duplication, should `ShiftedScaledDistribution` inherit from `IMFTransformedDistribution`, providing an override for `__init__` and any methods where knowing we're dealing with a linear transform let's us do better than the generic methods?

File: scipy/stats/_distribution_infrastructure.py
Comment (steppi) [reply scipy/stats/_distribution_infrastructure.py:5286]: This makes sense, yeah.

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [reply scipy/stats/_distribution_infrastructure.py:5222]: Maybe. `ShiftedScaledDistribution` also allows a negative scale, though (so we don't have to have separate right and left versions of distributions like `weibull_min`/`weibull_max` and `gumbel_l`/`gumbel_r`), so it is not strictly a monotonic transformation.

Comment (steppi) [reply scipy/stats/_distribution_infrastructure.py:5222]: Exactly, `increasing` changes depending on the sign of the scale. It seems like the principled way to do things here, but obviously shouldn't hold up anything.

File: scipy/stats/_distribution_infrastructure.py
Comment (steppi) [reply scipy/stats/_distribution_infrastructure.py:5248]: It's definitely better, but it would be good if it had enough info to recover what the transform was if we're making these public. It shouldn't be too much work to get things like `exp(Normal(mu=0.0, sigma=1.0))` and `1/Gamma(a=2)`.

File: scipy/stats/_distribution_infrastructure.py
Comment (steppi) [reply scipy/stats/_distribution_infrastructure.py:5222]: Oh right, and then things start to look more complicated.

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:4331]: Should probably raise if `np.any(X.support()[0) < 0`.

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [inline scipy/stats/_distribution_infrastructure.py:1842]: Should test these.

File: scipy/stats/_distribution_infrastructure.py
Comment (steppi) [inline scipy/stats/_distribution_infrastructure.py:4171]: I like how you did this.

**Review by steppi — APPROVED** [review]
I'm pretty happy with this. I personally don't think it's a problem to put things like `exp` and `log` for transforms of distributions / random variables directly in the `scipy.stats` namespace. If they had to go in a nested namespace, I think `scipy.stats.transforms`  could be a decent option.

@josef-pkt had some concerns about how fitting would work. I think the examples @mdhaber showed, where fresh instances are instantiated instead of mutating parameters of an existing instance, demonstrate that fitting without mutability is viable at least. In any case, though there currently isn't a public way to mutate the parameters of an instantiated distribution, if that ability was added to the public API, there's a question for how the parameters of derived distributions should behave. Currently a `MonotonicTransformedDistribution` carries a reference to the original distribution and any updates made to the original distribution would be reflected in the derived distribution. For example

```Python
import scipy.stats as stats
from scipy.stats import Normal

X = Normal(mu=0.0, sigma=1.0)
Y = stats.exp(X)
print(Y)
# exp(Normal(mu=0.0, sigma=1.0))
X._update_parameters(mu=1.0, sigma=2.0)
print(Y)
# exp(Normal(mu=1.0, sigma=2.0))
```

where we've used the private `_update_parameters` method which is used internally. As an implementation detail, updates in the parameters of `Y` above would not be reflected in `X`, and would not actually impact `Y` in any material way. If we open the door to mutability, we need to figure out how this stuff should behave. Do the parameters of `X` become first class parameters of `Y` which can be updated through `Y`. Should updates to `X` be reflected in `Y`; that sounds plausible under the transformations of random variables interpretation at least. Should updates to `Y` be reflected in `X`, or should `Y` be treated sort of like a view, modifiable only through `X`?  

If instances are treated as immutable for now, we don't have to answer these questions yet. If mutability is necessary to support `statsmodels`, I think it's reasonable to add it, and then just make sure we think through the semantics of the transformations and make sure everything behaves clearly and consistently.

@fancidev had some good thoughts on whether we should also support transforming a distribution class to get a new distribution class. I don't think it actually allows for anything that can't be done by instantiating first and then transforming though. Transforming the instances seems more natural to me too, so if we wanted to have one and only one way to do it, that would be my preference.

**Comment by tylerjereddy** [comment]
@fancidev @josef-pkt any thoughts here? is this ready to go before we branch for the `1.15.0` release in a few days?

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [reply scipy/stats/_distribution_infrastructure.py:4178]: Done.

**Comment by steppi** [comment]
If no one objects, I would like to merge this today so we can get it in for 1.15. A [discourse post](https://discuss.scientific-python.org/t/enh-stats-transformations-of-random-variables-and-more/1564) went up on Monday and no one has objected yet.

**Review by tupui — APPROVED** [review]

File: scipy/stats/_distribution_infrastructure.py
Comment (jorenham) [inline scipy/stats/_distribution_infrastructure.py:4176]: It could help with static typing if `dist` could be made positional-only (and also in the superclass). Otherwise I'd have to also account for the situation where there are no `*args` and `dist` is passed by keyword.

Comment (jorenham) [reply scipy/stats/_distribution_infrastructure.py:4176]: much appreciated

**Comment by steppi** [comment]
I'm going to go ahead and push the button on this one. We have two approvals, a thumbs up from @fancidev, and no objections. Thanks everyone

File: scipy/stats/_distribution_infrastructure.py
Comment (mdhaber) [reply scipy/stats/_distribution_infrastructure.py:4176]: Done in https://github.com/scipy/scipy/pull/21777/commits/47cff617e5530cdf0764a4c7e123fc66753bc378

**Comment by tylerjereddy** [comment]
This may need a release note if there isn't one already.

**Pull request merged**
